[WPE] Increased memory consumption of WebProcess during regular video playback and RWI

pgorszkowski-igalia opened the pull request:
#### d84062d7b5e2a29f7c29c1edf5bbfcf5eae6935b
<pre>
[WPE] Increased memory consumption of WebProcess during regular video playback and RWI
<a href="https://bugs.webkit.org/show_bug.cgi?id=284108">https://bugs.webkit.org/show_bug.cgi?id=284108</a>

Reviewed by Devin Rousso.

When WebInspector is used, the downloaded data is cached also for it in NetworkResourcesData.
The single resource data cannot be greater than maximumSingleResourceContentSize (50MB) and
the total size of resources content cannot be greater than maximumResourcesContentSize (200MB).
At the end of downloading the resource, the binary data is decoded to the string represantation
(base64 or other depends on decoder). Decoding process can increased the size of the kept resource
data. The limit maximumSingleResourceContentSize is checked but maximumResourcesContentSize limit
is not checked which can lead to situation that m_contentSize (the total size of resources content)
is greater than maximumResourcesContentSize. This causes that condition checked in
NetworkResourcesData::ensureFreeSpace is invalid because subtraction unsigned values where
first value(minuend) is smaller than the second one(subtrahend) gives a huge number (instead of
negative number).

This change ensures that after decoding binary data into string representation the total size of
resources content (m_contentSize) is not greater than maximumResourcesContentSize. The assert is
added in NetworkResourcesData::ensureFreeSpace to check that m_contentSize is not greater than
maximumResourcesContentSize.

I fixed implementation of function NetworkResourcesData::clear in case passing the preservedLoaderId.
In that case we should update m_requestIdsDeque and m_contentSize and not just clear them.

Additionally I changed the type of m_requestIdsDeque from Deque&lt;String&gt; to ListHashSet&lt;String&gt;.
This change fixes adding to m_requestIdsDeque the same requestId many times.

* Source/WebCore/inspector/NetworkResourcesData.cpp:
(WebCore::NetworkResourcesData::ResourceData::decodeDataToContent):
(WebCore::NetworkResourcesData::setResourceContent):
(WebCore::NetworkResourcesData::maybeAddResourceData):
(WebCore::NetworkResourcesData::maybeDecodeDataToContent):
(WebCore::NetworkResourcesData::clear):
(WebCore::NetworkResourcesData::ensureFreeSpace):
* Source/WebCore/inspector/NetworkResourcesData.h:

Canonical link: <a href="https://commits.webkit.org/289143@main">https://commits.webkit.org/289143@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/07fe1f6d527f3c0b2e2f8e8afabf49bfbb801087

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85217 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4952 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39648 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90344 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36257 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87306 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5041 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12927 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66254 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24071 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88263 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3825 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77401 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46562 "Found 2 new API test failures: /WPE/TestSSL:/webkit/WebKitWebView/web-socket-client-side-certificate, /WPE/TestSSL:/webkit/WebKitWebView/web-socket-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3701 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31662 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35325 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74460 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32495 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91789 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12564 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9150 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74792 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12793 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73238 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73909 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18342 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18311 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16736 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4551 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12507 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17960 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12337 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15830 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14087 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/inspector/NetworkResourcesData.cpp:
```diff
@@ -260,6 +260,13 @@ void NetworkResourcesData::maybeDecodeDataToContent(const String& requestId)
         m_contentSize -= resourceData->evictContent();
+
+    while (m_contentSize > m_maximumResourcesContentSize) {
+        String requestId = m_requestIdsDeque.takeFirst();
```
hmm im not really sure about this.  do we know for sure what the ordering of resources is inside of this?  i have a feeling that the first thing added is the `.html`, which we probably never want to evict

i think our current existing behavior is that if the developer requests enough things such that we run out of space, Web Inspector will show an error when trying to preview the resource content

i can definitely understand with and agree that that's not a great experience, but i think we need to be smarter about what things to evict from the cache

pgorszkowski-igalia replied:
This approach was taken by me base how it is done in `NetworkResourcesData::ensureFreeSpace`. If it is wrong we should fix that also there (separate PR?).

I have an idea for another PR: Do we need to gather data that cannot be displayed in the inspector (video or part of it)?

pgorszkowski-igalia replied:
@dcrousso : how about not removing from cache items with type InspectorPageAgent::DocumentResource as long as it is possible?

dcrousso replied:
hmmm maybe it's actually fine to get rid of the oldest resource since it should've already been sent to the frontend?

pgorszkowski-igalia replied:
hmm, actually if we have a site which loads a lot of resources and it exceeds the limit of total resources cache, than webkit removes the main document resource data from the cache and it will not be available in web inspector.

dcrousso commented on Source/WebCore/inspector/NetworkResourcesData.cpp:
```diff
@@ -256,10 +257,13 @@ void NetworkResourcesData::maybeDecodeDataToContent(const String& requestId)
 
-    m_contentSize += resourceData->decodeDataToContent();
-    size_t dataLength = resourceData->content().sizeInBytes();
-    if (dataLength > m_maximumSingleResourceContentSize)
-        m_contentSize -= resourceData->evictContent();
```
is this no longer useful/desirable?  id think that if after decoding the resource is too big we'd wanna get rid of it

pgorszkowski-igalia replied:
ah, you are right, it is checked also in ensureFreeSpace but it is not evicted in case the new size of resource is bigger than the limit. I will correct it.

dcrousso commented on Source/WebCore/inspector/NetworkResourcesData.cpp:
```diff
@@ -316,19 +325,24 @@ Vector<String> NetworkResourcesData::removeCachedResource(CachedResource* cached
+        m_requestIdsDeque.clear();
+        m_contentSize = 0;
+    } else {
+        ListHashSet<String> requestIdsDeque(WTFMove(m_requestIdsDeque));
```
NIT: id do this instead
```cpp
auto&& requestIdsDeque = std::exchange(m_requestIdsDeque, { });
```

dcrousso commented on Source/WebCore/inspector/NetworkResourcesData.cpp:
```diff
@@ -316,19 +325,24 @@ Vector<String> NetworkResourcesData::removeCachedResource(CachedResource* cached
+            if (resourceData->loaderId() == *preservedLoaderId)
+                m_requestIdsDeque.add(requestId);
+            else {
+                m_contentSize -= resourceData->evictContent();
```
rather than subtracting from `m_contentSize` i think it might be more accurate to recalculate `m_contentSize` from what's in `m_requestIdsDeque` after iterating (or perhaps when the `requestId` is added back to `m_requestIdsDeque`)

pgorszkowski-igalia replied:
If we want to recalculate the m_contentSize (after iterating or when the `requestId` is added back) then we need to get the size of the content or data kept in `resourceData` which is currently not available (there is no one function to get it, we need to check data or content to get the size of one of them). Function `evictContent` provides us that information that is, why I chose this approach.
Do you think it would be better to add a function which returns the current size of `resourceData`(data or content) and recalculate `m_contentSize`?

dcrousso commented on Source/WebCore/inspector/NetworkResourcesData.cpp:
```diff
@@ -316,19 +325,24 @@ Vector<String> NetworkResourcesData::removeCachedResource(CachedResource* cached
-            if (value->loaderId() != *preservedLoaderId)
-                keysToRemove.append(key);
+        m_requestIdsDeque.clear();
+        m_contentSize = 0;
```
Style: i'd make this an early return so that we can deindent all the rest of the logic for slightly nicer readability

dcrousso commented on Source/WebCore/inspector/NetworkResourcesData.cpp:
```diff
@@ -256,10 +257,18 @@ void NetworkResourcesData::maybeDecodeDataToContent(const String& requestId)
+    auto dataLength = resourceData->dataLength();
+    m_contentSize -= dataLength;
+
+    resourceData->decodeDataToContent();
```
can you explain more why the code that was there before is not good enough?

additionally, with this change does `NetworkResourcesData::ResourceData::decodeDataToContent` need to return a value anymore?  if not, can we adjust that function to remove that unnecessary logic?

pgorszkowski-igalia replied:
`ensureFreeSpace` bases on `m_contentSize` so that is why I had to adopt existing code to new condition `if (ensureFreeSpace(dataLength) && !resourceData->isContentEvicted())`. 
So now the code:
1. removes the size of data kept in `resourceData` from `m_contentSize`
2. decodes data to content (`decodeDataToContent`)
3. gets the size of content after decoding from data
4. ensures space for the size of content
5. if there is enough space and the content was not removed, `m_contentSize` is increased by content size.

You are right, we can remove returned value from `NetworkResourcesData::ResourceData::decodeDataToContent`.

dcrousso commented on Source/WebCore/inspector/NetworkResourcesData.cpp:
```diff
@@ -256,10 +256,18 @@ void NetworkResourcesData::maybeDecodeDataToContent(const String& requestId)
+    m_contentSize -= dataLength;
+
+    resourceData->decodeDataToContent();
+    dataLength = resourceData->content().sizeInBytes();
```
NIT: rather than reassign to `dataLength` could we instead have this be called `byteCount` or something?

dcrousso commented on Source/WebCore/inspector/NetworkResourcesData.cpp:
```diff
@@ -316,19 +324,25 @@ Vector<String> NetworkResourcesData::removeCachedResource(CachedResource* cached
+
+    auto&& requestIdsDeque = std::exchange(m_requestIdsDeque, { });
+
+    for (const auto& requestId : requestIdsDeque) {
```
NIT: id inline this
```cpp
for (auto&& requestId : std::exchange(m_requestIdsDeque, { })) {
```

dcrousso reviewed (approved):
nice work!  thanks for iterating with me :)

pgorszkowski-igalia committed:
[WPE] Increased memory consumption of WebProcess during regular video playback and RWI
https://bugs.webkit.org/show_bug.cgi?id=284108

Reviewed by Devin Rousso.

When WebInspector is used, the downloaded data is cached also for it in NetworkResourcesData.
The single resource data cannot be greater than maximumSingleResourceContentSize (50MB) and
the total size of resources content cannot be greater than maximumResourcesContentSize (200MB).
At the end of downloading the resource, the binary data is decoded to the string represantation
(base64 or other depends on decoder). Decoding process can increased the size of the kept resource
data. The limit maximumSingleResourceContentSize is checked but maximumResourcesContentSize limit
is not checked which can lead to situation that m_contentSize (the total size of resources content)
is greater than maximumResourcesContentSize. This causes that condition checked in
NetworkResourcesData::ensureFreeSpace is invalid because subtraction unsigned values where
first value(minuend) is smaller than the second one(subtrahend) gives a huge number (instead of
negative number).

This change ensures that after decoding binary data into string representation the total size of
resources content (m_contentSize) is not greater than maximumResourcesContentSize. The assert is
added in NetworkResourcesData::ensureFreeSpace to check that m_contentSize is not greater than
maximumResourcesContentSize.

I fixed implementation of function NetworkResourcesData::clear in case passing the preservedLoaderId.
In that case we should update m_requestIdsDeque and m_contentSize and not just clear them.

Additionally I changed the type of m_requestIdsDeque from Deque<String> to ListHashSet<String>.
This change fixes adding to m_requestIdsDeque the same requestId many times.

* Source/WebCore/inspector/NetworkResourcesData.cpp:
(WebCore::NetworkResourcesData::ResourceData::decodeDataToContent):
(WebCore::NetworkResourcesData::setResourceContent):
(WebCore::NetworkResourcesData::maybeAddResourceData):
(WebCore::NetworkResourcesData::maybeDecodeDataToContent):
(WebCore::NetworkResourcesData::clear):
(WebCore::NetworkResourcesData::ensureFreeSpace):
* Source/WebCore/inspector/NetworkResourcesData.h:

Canonical link: https://commits.webkit.org/289143@main

blino commented on Source/WebCore/inspector/NetworkResourcesData.cpp:
```diff
@@ -114,7 +114,7 @@ unsigned NetworkResourcesData::ResourceData::decodeDataToContent()
     }
 
-    return m_content.sizeInBytes() - buffer->size();
+    ASSERT(m_content.sizeInBytes() >= buffer->size());
```
@pgorszkowski-igalia : This assert is causing a crash for me on resources containing an UTF8 NBSP character (U+00A0, 0xC2 0xA0 sequence).
This is decoded to a single character by `decodeNonASCIISequence()` here: https://github.com/WebKit/WebKit/blob/main/Source/WebCore/PAL/pal/text/TextCodecUTF8.cpp#L100

pgorszkowski-igalia replied:
@blino : good point, thanks. My assumption was that after decoding, the content cannot be smaller, but it seems, that in some cases it can. In that case, I think, I can remove this assertion.

pgorszkowski-igalia replied:
@blino : can you provide an url on which you have this assert?

blino replied:
@pgorszkowski-igalia : I got the assert from this URL: https://www.youtube.com/s/_/kabuki/_/js/k=kabuki.base.en_US.DjiexSoKW-s.es5.O/ck=kabuki.base.LeEAEZDvUTs.L.W.O/am=AJBEIAAAAAoAEA/d=1/exm=HZgGtc,HpJhd,OoTvYc,base,main,mlDGgf/excm=main/ed=1/rs=ANjRhVluC6KmnfuWHJdF4GOkO58uR1_fTA/m=z1bKr,MBIy0b,ueuEUd,tPoMT,fBP14d,dWGIDd,sLflkd,hfnUMe,mTWISc,AZJeef,H4yxTb,YIRiad,ExM27,VPMWWc,aqohWb,QihACe,ld1uBb,LIz9J,NcwsS,TCNhee,iGSLje,g3Teq,JFqBjd,uQyaBf

It fails because of this content:
```
.yt-keyboard-key--disabled::before{content:\" \"}
```

pgorszkowski-igalia replied:
I created PR with fix for that: https://github.com/WebKit/WebKit/pull/39713
